toolchain/check/context.cpp

```diff
@@ -306,37 +383,66 @@ auto Context::LookupUnqualifiedName(Parse::NodeId node_id,
 auto Context::LookupNameInExactScope(SemIRLoc loc, SemIR::NameId name_id,
                                      SemIR::NameScopeId scope_id,
                                      const SemIR::NameScope& scope)
-    -> SemIR::InstId {
+    -> std::pair<SemIR::InstId, SemIR::AccessKind> {
   if (auto lookup = scope.name_map.Lookup(name_id)) {
-    auto inst_id = scope.names[lookup.value()].inst_id;
-    LoadImportRef(*this, inst_id);
-    return inst_id;
+    auto entry = scope.names[lookup.value()];
+    LoadImportRef(*this, entry.inst_id);
+    return {entry.inst_id, entry.access_kind};
   }
+
   if (!scope.import_ir_scopes.empty()) {
-    return ImportNameFromOtherPackage(*this, loc, scope_id,
-                                      scope.import_ir_scopes, name_id);
+    // TODO: Enforce other access modifiers for imports
+    return {ImportNameFromOtherPackage(*this, loc, scope_id,
+                                       scope.import_ir_scopes, name_id),
+            SemIR::AccessKind::Public};
   }
-  return SemIR::InstId::Invalid;
+  return {SemIR::InstId::Invalid, SemIR::AccessKind::Public};
 }
 
 auto Context::LookupQualifiedName(Parse::NodeId node_id, SemIR::NameId name_id,
-                                  LookupScope scope, bool required)
+                                  LookupScope scope, bool required,
+                                  std::optional<AccessInfo> access_info)
     -> LookupResult {
   llvm::SmallVector<LookupScope> scopes = {scope};
   LookupResult result = {.specific_id = SemIR::SpecificId::Invalid,
                          .inst_id = SemIR::InstId::Invalid};
   bool has_error = false;
+  bool is_parent_access = false;
+
+  auto is_access_allowed = [access_info](SemIR::AccessKind access_kind,
+                                         bool is_parent_access) -> bool {
+    if (!access_info || access_kind == SemIR::AccessKind::Public) {
+      return true;
+    }
+
+    if (access_kind == SemIR::AccessKind::Private &&
+        access_info->allow_private_access && !is_parent_access) {
+      return true;
+    }
+
+    if (access_kind == SemIR::AccessKind::Protected &&
+        access_info->allow_protected_access) {
+      return true;
+    }
+
+    return false;
```

Comment:
Had you considered writing this as a `switch(access_kind)` instead of individually checking `access_kind` values?

e.g.:

```suggestion
    if (!access_info) {
      return true;
    }
    switch (access_kind) {
      case SemIR::AccessKind::Public:
        return true;
      case SemIR::AccessKind::Protected:
        return access_info->allow_protected_access;
      case SemIR::AccessKind::Private:
        return access_info->allow_private_access && !is_parent_access;
    }
```